### PR TITLE
Add new custom type `Arg`, update API to `main! : List Arg => Result {} [Exit I32 Str]_`

### PR DESCRIPTION
**Pull request by lukewilliamboswell**

### CODE REVIEW

File: crates/roc_host_lib/src/lib.rs [hunk 1]
```diff
@@ -8,16 +7,19 @@ use std::borrow::Borrow;
...
+    let args = std::slice::from_raw_parts(argv, argc)
         .iter()
         .map(|&c_ptr| {
             let c_str = std::ffi::CStr::from_ptr(c_ptr);
-            let roc_str = RocStr::from(c_str.to_string_lossy().borrow());
-            ReadOnlyRocStr::from(roc_str)
+
+            // TODO confirm this is ok... feels dangerous
+            let os_str =
+                std::ffi::OsString::from_encoded_bytes_unchecked(c_str.to_bytes().to_owned());
```
Comment (lukewilliamboswell) [inline crates/roc_host_lib/src/lib.rs:17]: Please read the docs and give me an opinion here 🙏

File: platform/Arg.roc [hunk 2]
```diff
@@ -1,10 +1,42 @@
...
+import InternalArg
 
-## Gives a list of the program's command-line arguments.
-list! : {} => List Str
-list! = \{} ->
-    Host.args! {}
+# os_str's are not necessarily valid utf-8 or utf-16
+# so we store as raw bytes internally to avoid
+# common mistakes
+Arg := InternalArg.ArgToAndFromHost
```
Comment (smores56) [inline platform/Arg.roc:13]:
```suggestion
## An OS-aware representation of a command-line argument.
##
## Though we tend to think of args as Unicode strings, most operating systems
## represent command-line arguments as lists of bytes that aren't necessarily
## UTF-8 encoded. Windows doesn't even use bytes, but U16s.
##
## Most of the time, you will pass these to packages and they will handle the
## encoding for you, but for quick-and-dirty code you can use [display] to
## convert these to [Str] in a lossy way.
Arg := InternalArg.ArgToAndFromHost
```

File: platform/Arg.roc [hunk 3]
```diff
@@ -1,10 +1,42 @@
...
+to_os_raw : Arg -> [Unix (List U8), Windows (List U16)]
+to_os_raw = \@Arg inner ->
+    when inner.type is
+        Unix -> Unix inner.unix
+        Windows -> Windows inner.windows
+
+from_os_raw : InternalArg.ArgToAndFromHost -> Arg
+from_os_raw = \raw -> @Arg raw
+
+## Convert an Arg to a `Str` for display purposes
```
Comment (smores56) [inline platform/Arg.roc:26]:
```suggestion
## Convert an Arg to a `Str` for display purposes.
```

File: platform/Arg.roc [hunk 4]
```diff
@@ -1,10 +1,42 @@
...
+
+## Unwrap the raw bytes for decoding, typically this is
+## consumed by a package and not an end user
+to_os_raw : Arg -> [Unix (List U8), Windows (List U16)]
+to_os_raw = \@Arg inner ->
+    when inner.type is
+        Unix -> Unix inner.unix
+        Windows -> Windows inner.windows
+
+from_os_raw : InternalArg.ArgToAndFromHost -> Arg
```
Comment (smores56) [inline platform/Arg.roc:23]: Also, this should probably take `[Unix (List U8), Windows (List U16)]` to not expose internals.

File: crates/roc_host_lib/src/lib.rs (hunk 1 above)
Comment (smores56) [reply crates/roc_host_lib/src/lib.rs:17]: > The byte encoding is an unspecified, platform-specific, self-synchronizing superset of UTF-8. By being a self-synchronizing superset of UTF-8, this encoding is also a superset of 7-bit ASCII.

So I think this is incorrect, but it also seems unnecessary. This seems like we could just pass the `&[u8]` to `ArgToAndFromHost::from(arg)` and handle it there. For Unix it's handled, and for Windows we'll need to do something special anyway.

**Review by smores56 — CHANGES_REQUESTED** [review]

File: crates/roc_host_lib/src/lib.rs (hunk 1 above)
Comment (lukewilliamboswell) [reply crates/roc_host_lib/src/lib.rs:17]: fixed in f6680903c

File: examples/command.roc [hunk 5]
```diff
@@ -3,7 +3,7 @@ app [main!] { pf: platform "../platform/main.roc" }
 import pf.Stdout
 import pf.Cmd
 
-main! = \{} ->
+main! = \_ ->
```
[the remark this answers is not in the document]
Comment (lukewilliamboswell) [reply examples/command.roc:6]: updated in f5d933588

**Review by smores56 — APPROVED** [review]
Thanks for making the change!

**Pull request merged**